govcd/defined_interface_test.go

```diff
@@ -0,0 +1,127 @@
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	. "gopkg.in/check.v1"
+	"strings"
+)
+
+// Test_DefinedInterface tests the CRUD behavior of Defined Interfaces as a System administrator and tenant user.
+// This test can be run with GOVCD_SKIP_VAPP_CREATION option enabled.
+func (vcd *TestVCD) Test_DefinedInterface(check *C) {
+	if vcd.skipAdminTests {
+		check.Skip(fmt.Sprintf(TestRequiresSysAdminPrivileges, check.TestName()))
+	}
+	skipOpenApiEndpointTest(vcd, check, types.OpenApiPathVersion1_0_0+types.OpenApiEndpointInterfaces)
+	if len(vcd.config.Tenants) == 0 {
+		check.Skip("skipping as there is no configured tenant users")
+	}
+
+	// Creates the clients for the System admin and the Tenant user
+	systemAdministratorClient := vcd.client
+	tenantUserClient := NewVCDClient(vcd.client.Client.VCDHREF, true)
+	err := tenantUserClient.Authenticate(vcd.config.Tenants[0].User, vcd.config.Tenants[0].Password, vcd.config.Tenants[0].SysOrg)
+	check.Assert(err, IsNil)
+
+	// First, it checks how many exist already, as VCD contains some pre-defined ones.
+	allDefinedInterfacesBySysAdmin, err := systemAdministratorClient.GetAllDefinedInterfaces(nil)
+	check.Assert(err, IsNil)
+	alreadyPresentRDEs := len(allDefinedInterfacesBySysAdmin)
+
+	allDefinedInterfacesByTenant, err := tenantUserClient.GetAllDefinedInterfaces(nil)
+	check.Assert(err, IsNil)
+	check.Assert(len(allDefinedInterfacesByTenant), Equals, len(allDefinedInterfacesBySysAdmin))
+
+	// Then we create a new Defined Interface with System administrator. We replace the dots in both
+	// namespace and name as API is quirky at versions of VCD < 37.0
+	dummyRde := &types.DefinedInterface{
+		Name:      strings.ReplaceAll(check.TestName()+"name", ".", ""),
+		Namespace: strings.ReplaceAll(check.TestName()+"nss", ".", ""),
+		Version:   "1.2.3",
+		Vendor:    "vmware",
+	}
+	newDefinedInterfaceFromSysAdmin, err := systemAdministratorClient.CreateDefinedInterface(dummyRde)
+	check.Assert(err, IsNil)
+	check.Assert(newDefinedInterfaceFromSysAdmin, NotNil)
+	check.Assert(newDefinedInterfaceFromSysAdmin.DefinedInterface.Name, Equals, dummyRde.Name)
+	check.Assert(newDefinedInterfaceFromSysAdmin.DefinedInterface.Namespace, Equals, dummyRde.Namespace)
+	check.Assert(newDefinedInterfaceFromSysAdmin.DefinedInterface.Version, Equals, dummyRde.Version)
+	check.Assert(newDefinedInterfaceFromSysAdmin.DefinedInterface.Vendor, Equals, dummyRde.Vendor)
+	check.Assert(newDefinedInterfaceFromSysAdmin.DefinedInterface.IsReadOnly, Equals, dummyRde.IsReadOnly)
+	AddToCleanupListOpenApi(newDefinedInterfaceFromSysAdmin.DefinedInterface.ID, check.TestName(), types.OpenApiPathVersion1_0_0+types.OpenApiEndpointInterfaces+newDefinedInterfaceFromSysAdmin.DefinedInterface.ID)
+
+	// Tenants can't create Defined Interfaces. We replace the dots in both
+	// namespace and name as API is quirky at versions of VCD < 37.0
```

Comment:
Change `VCD < 37.0` to a specific VCD version instead, as the issue is bounded to VCD and not the API.
(ie, v36.0 in VCD 10.3.0, a namespace with dots is quirky, in VCD 10.4.0 the same version works well)